util/gateway/gateway.go

```diff
@@ -0,0 +1,167 @@
+package gateway
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"time"
+
+	"github.com/ipfs/go-blockservice"
+	"github.com/ipfs/go-cid"
+	blockstore "github.com/ipfs/go-ipfs-blockstore"
+	ipld "github.com/ipfs/go-ipld-format"
+	"github.com/ipfs/go-merkledag"
+	unixfs "github.com/ipfs/go-unixfs"
+	uio "github.com/ipfs/go-unixfs/io"
+)
+
+type GatewayHandler struct {
+	bs    blockstore.Blockstore
+	dserv ipld.DAGService
+}
+
+type httpError struct {
+	Code    int
+	Message string
+}
+
+func NewGatewayHandler(bs blockstore.Blockstore) *GatewayHandler {
+	return &GatewayHandler{
+		bs:    bs,
+		dserv: merkledag.NewDAGService(blockservice.New(bs, nil)),
+	}
+}
+
+func (gw *GatewayHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if err := gw.handleRequest(r.Context(), w, r); err != nil {
+		w.Header().Set("Content-Type", "application/json")
```

Comment:
I am pretty sure GW doesn't do that / and in general a random `curl`-er won't expect crap in their STDIN
Set the error as a header, and return an empty body?

Reply:
the main gateway uses http.Error, I can switch to that no problem